### DISCUSSION TITLE
Seeing multiple "reaped unknown child process" messages

### DISCUSSION THREAD

**Discussion by anson2590** [puma/puma#3420] [Q&A] [answered]
When we run application we are seeing multiple logs like "! reaped unknown child process pid=34 status=pid 34 exit 0" is there a way we can disable this log in our application ?

**To Reproduce**
- When the application runs on docker in splunk logs we are seeing lot of above of mentioned error message.

**Comment by dentarg** [comment]
It comes from  https://github.com/puma/puma/blob/a7e3d944fa66bbba355abb9efed8ce53048c7e7c/lib/puma/cluster.rb#L587

It was added in https://github.com/puma/puma/pull/3255

You can't disable this logging

**Comment by dentarg** [comment]
I think you want to know that this is happening and adjust your application for it to not happen. @byroot what do you think?

**Comment by byroot** [comment, marked as answer]
Agreed. This mean your application is spawning sub processes but not reaping them, which means you likely have lots of zombie processes.

I'm not against adding a way to disable that log, but it will just hide the problem, not fix it. I'd suggest to at least have a look into what these processes are.

**Comment by dentarg** [comment]
Thanks Jean. I'll convert this to a discussion for now.

**Comment by anson2590** [comment]
Thanks for quick response. Trying to understand if there is a zombie process should we kill them?

**Reply by dentarg** [reply]
https://en.wikipedia.org/wiki/Zombie_process

**Comment by anson2590** [comment]
Sorry this might be a wrong question I am trying to reproduce this log error running "rails server" instead of running in docker so that it will be easier to debug can you please help me how can I reproduce this issue.

in config/puma.rb I have this config:
`max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }`
`min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }`
`threads min_threads_count, max_threads_count`
`if ENV["RAILS_ENV"] == "development"`
`  require "concurrent-ruby"`
`  worker_count = Integer(ENV.fetch("WEB_CONCURRENCY") { Concurrent.physical_processor_count })`
`  workers worker_count if worker_count > 1`
`end`
`worker_timeout 3600 if ENV.fetch("RAILS_ENV", "development") == "development"`
`port ENV.fetch("PORT") { 3000 }`
`environment ENV.fetch("RAILS_ENV") { "development" }`
`pidfile ENV.fetch("PIDFILE") { "tmp/pids/server.pid" }`
`plugin :tmp_restart`

**Reply by byroot** [reply]
> Running "rails server" instead of running in docker so that it will be easier to debug

You can't. What makes docker special is that it spawn Puma as PID 1, on Unix PID 1 has special responsabilities, when a process is orphaned (daemonized) it is reparented to PID 1, and PID 1 is responsible for reaping it.

**Reply by byroot** [reply]
If you want to debug this further, one way would be to lauch puma in docker in such a way that it's not PID 1, e.g.: use a simple shell like: `sh -c "puma ..."`.

Then enter the container and look for zombie processes with `ps aux`.

**Reply by anson2590** [reply]
Sure Thanks, will try the suggested approach

**Comment by anson2590** [comment]
@byroot we have scenario where we make multiple service calls these calls will keep on running on backend parallel for infinite times that is until the server is stopped. This logic is put within the thread as this process keep running in background we are these reaped log. 
We have a question is it possible to put this log behind a env variable or some way that consuming app can disable this log if not required?

**Reply by byroot** [reply]
That's a question for the maintainers. But if these processes stay alive until the server shutdown, is this log really a problem?

All it is saying is that some sub process that isn't a puma worker was reparented and reaped. If you do expect that, it's fine to ignore. It's here to alert you in case you had no idea.

**Reply by anson2590** [reply]
Ok @byroot what these services does is it will keep on monitoring and report the application performance to another service. These code is implemented within an gem and it is consumed in our 50+ applications as it has consumed those many applications in splunk there are lot of logs keep showing up.

**Reply by byroot** [reply]
Honestly, if you don't want Puma to handle this, just use https://github.com/krallin/tini
